### PR TITLE
No xss in filters sidebar

### LINKED ISSUE #5198: filter behavior creates cross-site scripting (XSS) vulnerability
Found this old issue which calls out the problem, but only solves it for one specific case (and does so client-side using jQuery - which is probably not the most secure approach), I think: https://github.com/activeadmin/activeadmin/pull/3811

Here are the places in the code where user input is being rendered to the page without being sanitized:

https://github.com/activeadmin/activeadmin/blob/85d4690f824825fcd5f855146a56b7f98bcef37f/lib/active_admin/filters/active_sidebar.rb#L31

https://github.com/activeadmin/activeadmin/blob/8b15ff58a6d912899b46bf651354666d39f39ad0/lib/active_admin/batch_actions/views/batch_action_form.rb#L18

https://github.com/activeadmin/activeadmin/blob/9d0557ad05ece0da59a4df2c61468479093686a1/lib/active_admin/view_helpers/form_helper.rb#L14

### Expected behavior

What do you think should happen?

I should not be able to inject a `<script>` tag that will be executed when ActiveAdmin renders a page

### Actual behavior

What actually happens?

I can inject whatever I want because text input is not sanitized and then stored directly in `params[:q]`

### How to reproduce

- I enter `<script>alert('hax')</script>` in a text input in the filter sidebar
- I click filter
- I see an alert box with the text 'hax'
- I see a text input in the filter sidebar with value `<script>alert('hax')</script>`

(because this is currently the default behavior of ActiveAdmin, a reproduction repository should not be necessary)

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
Fixes #5198.
Supersedes #5265.

### CODE REVIEW

File: spec/unit/pretty_format_spec.rb [hunk 1]
```diff
@@ -8,12 +8,21 @@ def method_missing(*args, &block)
...
 
-  ['hello', 23, 5.67, 10**30, :foo, Arbre::Element.new.br(:foo)].each do |obj|
+  ['hello', 23, 5.67, 10**30, :foo].each do |obj|
     it "should call `to_s` on #{obj.class}s" do
       expect(pretty_format(obj)).to eq obj.to_s
     end
   end
 
+  it "normalizes Arbre elements" do
+    expect(pretty_format(Arbre::Element.new.br)).to eq("<br>\n")
```
Comment (deivid-rodriguez) [inline spec/unit/pretty_format_spec.rb]: This is an unexpected change in output from this patch. Before the ouput would be `<br />`, now it's `<br>`. Seems ok?

Comment (zorab47) [reply spec/unit/pretty_format_spec.rb]: https://stackoverflow.com/questions/1946426/html-5-is-it-br-br-or-br#1946446

For more information.

**Comment by zorab47** [comment]
Should it only be sanitize if the string isn't already consider `.html_safe?`? Perhaps users would optionally like to output HTML?

**Comment by deivid-rodriguez** [comment]
Mmm, I think that's what `sanitize` does? It cleans up some backlisted tags but allows the ones considered harmless. That's why `<br>` passes through, but `<script>` disappears in the `pretty_format` unit tests.

File: spec/unit/pretty_format_spec.rb (hunk 1 above)
Comment (javierjulio) [reply spec/unit/pretty_format_spec.rb]: Its fine. No need to be self closing in HTML5.

Comment (deivid-rodriguez) [reply spec/unit/pretty_format_spec.rb]: Only issue would be user searching for literal `<br />` on a text string but current filter will show `<br>`... Could be confusing somehow but seems reeeeally edge...

Comment (deivid-rodriguez) [reply spec/unit/pretty_format_spec.rb]: And an improvement anyways because previously nothing would be displayed, just some whitespace.

**Comment by markstory** [comment]
There is still an issue in the `else` case when an `ActiveRecord::Base` instance is provided.
The else case uses ends up in `auto_link` or in `display_name`

https://github.com/activeadmin/activeadmin/blob/a473cb68baee260021957ba3d6c3398459a4c551/lib/active_admin/view_helpers/display_helper.rb#L72-L79

The `auto_link` method  uses `display_name` as a default value,  https://github.com/activeadmin/activeadmin/blob/8b15ff58a6d912899b46bf651354666d39f39ad0/lib/active_admin/view_helpers/auto_link_helper.rb#L14

The current code for `display_name` is missing escaping 

https://github.com/activeadmin/activeadmin/blob/a473cb68baee260021957ba3d6c3398459a4c551/lib/active_admin/view_helpers/display_helper.rb#L17-L36

This an acute problem when the ActiveRecord model implements one of the `name` methods defined in 

https://github.com/activeadmin/activeadmin/blob/26b7c841073cdddbdc8307ae821711654d4ac2c4/lib/active_admin/application_settings.rb#L27-L34

The end result is XSS from user supplied data stored in active record models.

**Comment by deivid-rodriguez** [comment]
Thanks so much @markstory! Can you send us a patch fixing the other issues?

**Comment by markstory** [comment]
@deivid-rodriguez Sure, would you like a pull request to this branch or to master?

**Comment by deivid-rodriguez** [comment]
I think master is fine, it should be pretty independent from the code in this PR, right? Thanks!

**Comment by rdlugosz** [comment]
Yes, this seems heavy handed. In our project, this strips the `target=` attribute out of links, so you cannot have certain links open in new tabs. This even happens when a string is marked as `html_safe`, (for example, if you wrap the link_to in a call to raw like: `raw(link_to "foo", "bar")`).

I'm all for sanitizing the strings, but perhaps the step should be ignored when strings are declared to be safe?

**Comment by deivid-rodriguez** [comment]
Sorry about that, fixes welcome :pray:!

Or if the old situation is preferred, we can always revert my patch and release a new patch level version. I'm sorry I don't have time to contribute again now... :disappointed:

**Pull request merged**